src/main/java/com/amazonaws/gurureviewercli/util/RecommendationsFilter.java

```diff
@@ -0,0 +1,106 @@
+package com.amazonaws.gurureviewercli.util;
+
+import java.nio.file.FileSystems;
+import java.nio.file.PathMatcher;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import lombok.val;
+import software.amazon.awssdk.services.codegurureviewer.model.RecommendationSummary;
+
+import com.amazonaws.gurureviewercli.model.configfile.CustomConfiguration;
+
+/**
+ * Utility class to filter CodeGuru Reviewer recommendations based on the criteria in a
+ * custom configuration file.
+ */
+public final class RecommendationsFilter {
+
+    private static final String GLOB_PREFIX = "glob:";
+
+    private RecommendationsFilter() {
+        // do not instantiate.
+    }
+
+    /**
+     * Filter excluded recommendations.
+     *
+     * @param recommendations List of recommendations.
+     * @param configuration   Custom Configuration file that defines filters.
+     * @return Filtered list.
+     */
+    public static List<RecommendationSummary> filterRecommendations(
+        final Collection<RecommendationSummary> recommendations,
+        final CustomConfiguration configuration) {
+
+        val matchers = new ArrayList<PathMatcher>();
+        if (configuration.getExcludeFiles() != null) {
+            for (val globString : configuration.getExcludeFiles()) {
+                val matcher = FileSystems.getDefault().getPathMatcher(GLOB_PREFIX + globString);
+                matchers.add(matcher);
+            }
+        }
+
+        val result = new ArrayList<RecommendationSummary>();
+        for (val rec : recommendations) {
+            if (configuration.getExcludeBelowSeverity() != null) {
+                val threshold = RecommendationPrinter.severityToInt(configuration.getExcludeBelowSeverity());
+                if (RecommendationPrinter.severityToInt(rec) > threshold) {
+                    continue;
+                }
+            }
+            if (configuration.getExcludeById() != null &&
+                configuration.getExcludeById()
+                             .stream()
+                             .anyMatch(id -> id.equals(rec.recommendationId()))) {
+                continue;
+            }
+
+            if (matchers.stream().anyMatch(m -> m.matches(Paths.get(rec.filePath())))) {
+                continue;
+            }
+            if (rec.ruleMetadata() == null || rec.filePath().equals(".")) {
+                continue; // Always drop rules without metadata or the stats recomendation
```

Comment:
Could we add an option to keep the recommendations that don't have any metadata? I think it's great to drop them, but until a better mechanism exists, this will hide the problem under the rug. With such an option we can still enable it in end-to-end regression testing and alert.

Reply:
I'd rather fix this by making sure all rules have metadata. If there is no metadata, we don't know what rule it is, so we can't filter it. Usually, these are low severity, so customers who want to block pipelines using the CLI would likely suppress them anyway.

Reply:
Oh, and this only happens when you filter. So we can still get these by running without filter. OK, that's good enough.